lib/Dialect/Arc/Transforms/FindInitialVectors.cpp

```diff
@@ -0,0 +1,222 @@
+  unsigned get(Operation *op) const {
+    auto it = opRanks.find(op);
+    assert(it != opRanks.end() && "op has no rank");
+    return it->second;
+  }
+};
+} // namespace
+
+/// Assign each operation in the given block a topological rank. Stateful
+/// elements are assigned rank 0. All other operations receive the maximum rank
+/// of their users, plus one.
+LogicalResult TopologicalOrder::compute(Block *block) {
+  LLVM_DEBUG(llvm::dbgs() << "- Computing topological order in block " << block
+                          << "\n");
+  struct WorklistItem {
+    WorklistItem(Operation *op) : userIt(op->user_begin()) {}
+    Operation::user_iterator userIt;
+    unsigned rank = 0;
+  };
+  SmallMapVector<Operation *, WorklistItem, 16> worklist;
+  for (auto &op : *block) {
+    if (opRanks.contains(&op))
+      continue;
+    worklist.insert({&op, WorklistItem(&op)});
+    while (!worklist.empty()) {
+      auto &[op, item] = worklist.back();
+      if (auto stateOp = dyn_cast<StateOp>(op)) {
+        if (stateOp.getLatency() > 0)
+          item.userIt = op->user_end();
+      } else if (auto writeOp = dyn_cast<MemoryWritePortOp>(op)) {
+        item.userIt = op->user_end();
+      }
+      if (item.userIt == op->user_end()) {
+        opRanks.insert({op, item.rank});
+        worklist.pop_back();
+        continue;
+      }
+      if (auto rankIt = opRanks.find(*item.userIt); rankIt != opRanks.end()) {
+        item.rank = std::max(item.rank, rankIt->second + 1);
+        ++item.userIt;
+        continue;
+      }
+      if (!worklist.insert({*item.userIt, WorklistItem(*item.userIt)}).second)
+        return op->emitError("dependency cycle");
+    }
+  }
+  return success();
+}
+
+namespace {
+template <typename T>
+LogicalResult modifyKeyIfPossible(const T &container, std::string &key) {
+  if(container.empty())
+    return success();
+
+  if constexpr (std::is_same_v<T, StringAttr>) {
+    // The attributes' order doesn't matter so we need to sort the attributes
+    // in some order and the concatinate them
+    std::vector tempAttributeHolder(container.begin(), container.end());
+    std::sort(tempAttributeHolder.begin(), tempAttributeHolder.end());
+    key += std::accumulate(tempAttributeHolder.begin(), tempAttributeHolder.end(), "");
+    return success();
+  }
+
+  // OperandTypes or ResultTypes
+  for (const auto &elem : container) {
+    std::string str;
+    llvm::raw_string_ostream os(str);
+    elem.print(os);
+    key += str;
+  }
+  return success();
+}
+
+std::string computeKey(Operation *op) {
+  // The key = concat(op_rank, op_name, op_operands_types, op_result_types, op_attrs)
+  std::string key = op->getName().getStringRef().str();
+  if (failed(modifyKeyIfPossible(op->getOperandTypes(), key)) ||
+      failed(modifyKeyIfPossible(op->getResultTypes(), key)) || 
+      failed(modifyKeyIfPossible(op->getRegisteredInfo()->getAttributeNames(), key)))
+    assert(false && "Cannot compute the op key");
+
+  return key;
+}
```

Comment:
I would avoid using a string as key here. String representations can be pretty expensive to compute and are mostly intended for humans to read. You should be able to use an `std::tuple` instead:

```c++
using Key = std::tuple<...>;

Key computeKey(Operation *op, unsigned rank) {
  return std::make_tuple(
    rank,
    op->getName(),
    op->getOperandTypes(), // maybe SmallVector<2>(op->getOperandTypes())
    op->getResultTypes(), // maybe also in a SmallVector
    op->getAttributes()
  );
}
```

Reply:
```
using Key = std::tuple<unsigned, StringRef, SmallVector<Type, 4>,
                       SmallVector<Type, 4>, ArrayRef<NamedAttribute>>;

Key computeKey(Operation *op, unsigned rank) {
  SmallVector<Type, 4> operandTypes(op->operand_type_begin(), op->operand_type_end());
  SmallVector<Type, 4> resultTypes(op->result_type_begin(), op->result_type_end());
  // The key = concat(op_rank, op_name, op_operands_types, op_result_types, op_attrs)
  return std::make_tuple(rank, op->getName(), operandTypes, resultTypes, op->getAttrs());
}
```
```
error: no viable conversion from returned value of type 'tuple<[...], typename __decay_and_strip<OperationName>::__type, [3 * ...]>' to function return type 'tuple<[...], StringRef, [3 * ...]>'
  141 |   return std::make_tuple(rank, op->getName(), operandTypes, resultTypes, op->getAttrs());
```

any hint?